osu.Game.Rulesets.Taiko/Mods/TaikoModAlternate.cs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Linq;
+using osu.Framework.Bindables;
+using osu.Game.Configuration;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Objects.Types;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Rulesets.Taiko.Objects;
+
+namespace osu.Game.Rulesets.Taiko.Mods
+{
+    public class TaikoModAlternate : ModAlternate<TaikoHitObject, TaikoAction>
+    {
+        [SettingSource("Playstyle", "Preferred Alternate Playstyle")]
+        public Bindable<Playstyle> Style { get; } = new Bindable<Playstyle>();
+
+        private TaikoAction? lastAction1;
+        private TaikoAction? lastAction2;
+        private TaikoAction lastAction;
```

Comment:
These need an explanation at worst and a proper rename at best. Having three `lastAction`s with numbers in them is quite unreadable.

If I were to write something that doesn't necessitate these two, I'd go for:

```csharp
private bool shouldBlock(TaikoAction currentAction)
{
    var actionsToCheck = (lastAction, currentAction);

    switch (Style.Value)
    {
        case Playstyle.KDDK:
            return compareUnordered((TaikoAction.LeftCentre, TaikoAction.LeftRim), actionsToCheck)
                   || compareUnordered((TaikoAction.RightCentre, TaikoAction.RightRim), actionsToCheck);

        case Playstyle.KKDD:
            return compareUnordered((TaikoAction.LeftCentre, TaikoAction.RightCentre), actionsToCheck)
                   || compareUnordered((TaikoAction.LeftRim, TaikoAction.RightRim), actionsToCheck);

        default:
            throw new ArgumentOutOfRangeException(nameof(currentAction));
    }
}

private bool compareUnordered((TaikoAction, TaikoAction) expected, (TaikoAction, TaikoAction) actual)
    => (expected.Item1 == actual.Item1 && expected.Item2 == actual.Item2)
       || (expected.Item1 == actual.Item2 && expected.Item2 == actual.Item1);
```

I *think* this works but haven't tested. Buyer beware. (Mostly posting for the overall structure.)